### PR TITLE
Fix multi-threaded gtests

### LINKED ISSUE #9146: Test failure with column split
This was originally reported in https://buildkite.com/xgboost/xgboost-ci/builds/2386, but I was able to reproduce it with `GrowHistMaker.ColumnSplit`:

```bash
for i in {1..2000}; do echo $i; ./testxgboost --gtest_filter=GrowHistMaker.ColumnSplit || break; done
```

On average it fails after a few hundred runs:
```console
/home/rongou/src/xgboost/tests/cpp/tree/test_histmaker.cc:102: Failure
Expected equality of these values:
  json
    Which is: "  { \"nodeid\": 0, \"depth\": 0, \"split\": \"1\", \"split_condition\": 0.269745648, \"yes\": 1, \"no\": 2, \"missing\": 2 , \"children\": [\n    { \"nodeid\": 1, \"depth\": 1, \"split\": \"6\", \"split_condition\": 1.82024896, \"yes\": 3, \"no\": 4, \"missing\": 4 , \"children\": [\n      { \"nodeid\": 3, \"leaf\": -0.101172648 }, \n      { \"nodeid\": 4, \"leaf\": -0.00866571628 }\n    ]}, \n    { \"nodeid\": 2, \"depth\": 1, \"split\": \"8\", \"split_condition\": 0.795054197, \"yes\": 5, \"no\": 6, \"missing\": 5 , \"children\": [\n      { \"nodeid\": 5, \"depth\": 2, \"split\": \"1\", \"split_condition\": 0.406772137, \"yes\": 7, \"no\": 8, \"missing\": 8 , \"children\": [\n        { \"nodeid\": 7, \"leaf\": -0.167432114 }, \n        { \"nodeid\": 8, \"leaf\": -0.38392964 }\n      ]}, \n      { \"nodeid\": 6, \"depth\": 2, \"split\": \"6\", \"split_condition\": 0.456804186, \"yes\": 9, \"no\": 10, \"missing\": 10 , \"children\": [\n        { \"nodeid\": 9, \"leaf\": -0.152618349 }, \n        { \"nodeid\": 10, \"leaf\": 0.0083628241 }\n      ]}\n    ]}\n  ]}"
  expected_json
    Which is: "  { \"nodeid\": 0, \"depth\": 0, \"split\": \"1\", \"split_condition\": 0.269745648, \"yes\": 1, \"no\": 2, \"missing\": 2 , \"children\": [\n    { \"nodeid\": 1, \"depth\": 1, \"split\": \"6\", \"split_condition\": 1.82024896, \"yes\": 3, \"no\": 4, \"missing\": 4 , \"children\": [\n      { \"nodeid\": 3, \"leaf\": -0.101172648 }, \n      { \"nodeid\": 4, \"leaf\": -0.00866571628 }\n    ]}, \n    { \"nodeid\": 2, \"depth\": 1, \"split\": \"8\", \"split_condition\": 0.795054197, \"yes\": 5, \"no\": 6, \"missing\": 5 , \"children\": [\n      { \"nodeid\": 5, \"depth\": 2, \"split\": \"1\", \"split_condition\": 0.406772137, \"yes\": 7, \"no\": 8, \"missing\": 8 , \"children\": [\n        { \"nodeid\": 7, \"leaf\": -0.167432114 }, \n        { \"nodeid\": 8, \"leaf\": -0.38392964 }\n      ]}, \n      { \"nodeid\": 6, \"depth\": 2, \"split\": \"9\", \"split_condition\": 0.979574144, \"yes\": 9, \"no\": 10, \"missing\": 10 , \"children\": [\n        { \"nodeid\": 9, \"leaf\": -0.0286615752 }, \n        { \"nodeid\": 10, \"leaf\": -0.10902027 }\n      ]}\n    ]}\n  ]}"
With diff:
@@ -9,7 +9,7 @@
         { \"nodeid\": 8, \"leaf\": -0.38392964 }
       ]}, 
-      { \"nodeid\": 6, \"depth\": 2, \"split\": \"6\", \"split_condition\": 0.456804186, \"yes\": 9, \"no\": 10, \"missing\": 10 , \"children\": [
-        { \"nodeid\": 9, \"leaf\": -0.152618349 }, 
-        { \"nodeid\": 10, \"leaf\": 0.0083628241 }
+      { \"nodeid\": 6, \"depth\": 2, \"split\": \"9\", \"split_condition\": 0.979574144, \"yes\": 9, \"no\": 10, \"missing\": 10 , \"children\": [
+        { \"nodeid\": 9, \"leaf\": -0.0286615752 }, 
+        { \"nodeid\": 10, \"leaf\": -0.10902027 }
       ]}
     ]}
```

Not sure about the root cause.

### PR DESCRIPTION
**Pull request by rongou**
Turns out OpenMP doesn't play nice with C++ `std::thread`s, causing occasional test failures.

Fixes #9146

### CODE REVIEW

**Comment by rongou** [comment]
@trivialfis

File: tests/cpp/tree/test_histmaker.cc [hunk 1]
```diff
@@ -90,13 +90,16 @@ void TestColumnSplit(int32_t rows, bst_feature_t cols, RegTree const& expected_t
...
+  ASSERT_EQ(tree.NumExtraNodes(), 10);
+  ASSERT_EQ(tree[0].SplitIndex(), 1);
 
-  EXPECT_NE(tree[tree[0].LeftChild()].SplitIndex(), 0);
-  EXPECT_NE(tree[tree[0].RightChild()].SplitIndex(), 0);
+  ASSERT_NE(tree[tree[0].LeftChild()].SplitIndex(), 0);
+  ASSERT_NE(tree[tree[0].RightChild()].SplitIndex(), 0);
 
-  EXPECT_EQ(tree, expected_tree);
+  FeatureMap fmap;
```
Comment (trivialfis) [inline tests/cpp/tree/test_histmaker.cc:99]: Hmm, is there any difference?

File: src/data/simple_dmatrix.cc [hunk 2]
```diff
@@ -166,7 +166,7 @@ BatchSet<GHistIndexMatrix> SimpleDMatrix::GetGradientIndex(Context const* ctx,
   }
   if (!gradient_index_ || detail::RegenGHist(batch_param_, param)) {
     // GIDX page doesn't exist, generate it
-    LOG(INFO) << "Generating new Gradient Index.";
```
Comment (trivialfis) [inline src/data/simple_dmatrix.cc:169]: Is there an issue with info log?

Comment (rongou) [reply src/data/simple_dmatrix.cc:169]: It always gets printed in the test, which makes it harder to see when you are running the same test 1000 times. Is there another way to turn it off?

File: tests/cpp/tree/test_histmaker.cc (hunk 1 above)
Comment (rongou) [reply tests/cpp/tree/test_histmaker.cc:99]: Compare the JSON string instead of the binary blobs.

**Review by trivialfis — APPROVED** [review]

**Pull request merged**